File: backend/app/api/exports.py
"""课表导出:班级、教师和教室/场地 Excel/PDF/PNG、全校总表、批量 zip(M5-1)。

Excel 在 api 同步生成(openpyxl 轻量);PDF/PNG 派到 worker(WeasyPrint + 中文字体)
再取回。单一对象导出开放给所有登录者(课表本就全校可查);全校总表/批量限排课管理员以上。
"""

from urllib.parse import quote

from fastapi import APIRouter, Depends, HTTPException, Query, Response, status
from sqlalchemy.orm import Session

from app.core.auth import get_active_user, require_roles
from app.core.db import get_db
from app.models.user import Role, User
from app.services import school_rules
from app.services import timetable_export as tex
from app.workers import queue as job_queue

router = APIRouter(tags=["exports"])

manager = require_roles(Role.scheduler, Role.director)

_MIME = {
    "xlsx": "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
    "pdf": "application/pdf",
    "png": "image/png",
    "zip": "application/zip",
}


def _download(data: bytes, filename: str, ext: str) -> Response:
    # 中文文件名以 RFC 5987 filename* 表达;filename 提供 ASCII 后备
    quoted = quote(filename)
    disposition = f"attachment; filename=\"export.{ext}\"; filename*=UTF-8''{quoted}.{ext}"
    return Response(
        content=data,
        media_type=_MIME[ext],
        headers={"Content-Disposition": disposition},
    )


@router.get("/export/timetable")
def export_timetable(
    semester_id: int = Query(...),
    view: str = Query(..., pattern="^(class|teacher|room)$"),
    target_id: int = Query(...),
    fmt: str = Query("xlsx", pattern="^(xlsx|pdf|png)$"),
    db: Session = Depends(get_db),
    _: User = Depends(get_active_user),
):
    """单一班级、教师或教室/场地的课表,格式 xlsx/pdf/png。"""
    try:
        grid, meta = tex.build_grid(db, semester_id, view, target_id)
    except tex.ExportError as e:
        raise HTTPException(status.HTTP_404_NOT_FOUND, str(e)) from e

    filename = f"{grid.title}_{meta.semester_label}"
    if fmt == "xlsx":
        return _download(tex.grids_to_xlsx([grid], meta), filename, "xlsx")

    html = tex.grid_to_html(grid, meta)
    try:
        data = job_queue.render_export(html, fmt)
    except job_queue.RenderError as e:
        raise HTTPException(status.HTTP_502_BAD_GATEWAY, str(e)) from e
    return _download(data, filename, fmt)


@router.get("/export/school.xlsx")
def export_school(
    semester_id: int = Query(...),
    db: Session = Depends(get_db),
    _: User = Depends(manager),
):
    """全校总表:一个 Excel,每班一个分页。"""
    try:
        data = tex.school_workbook(db, semester_id)
    except tex.ExportError as e:
        raise HTTPException(status.HTTP_404_NOT_FOUND, str(e)) from e
    return _download(data, f"{school_rules.export_label('school_timetable')}_{semester_id}", "xlsx")


@router.get("/export/batch.zip")
def export_batch(
    semester_id: int = Query(...),
    db: Session = Depends(get_db),
    _: User = Depends(manager),
):
    """批量导出:全部班级各一个 Excel,打包成 zip。"""
    try:
        data = tex.class_batch_zip(db, semester_id)
    except tex.ExportError as e:
        raise HTTPException(status.HTTP_404_NOT_FOUND, str(e)) from e
    return _download(data, f"{school_rules.export_label('class_timetables')}_{semester_id}", "zip")

File: frontend/src/api/exports.ts
// 课表导出下载(M5-1)。以 fetch 取回 blob,好处理加载状态、错误与中文文件名。

export type ExportFmt = 'xlsx' | 'pdf' | 'png'
export type ExportView = 'class' | 'teacher' | 'room'

function filenameFrom(cd: string | null, fallback: string): string {
  const m = cd ? /filename\*=UTF-8''([^;]+)/.exec(cd) : null
  return m ? decodeURIComponent(m[1]) : fallback
}

async function download(path: string, fallback: string): Promise<void> {
  const resp = await fetch(`/api${path}`, { credentials: 'include' })
  if (!resp.ok) {
    let detail: string | undefined
    try { detail = (await resp.json())?.detail } catch { detail = undefined }
    throw new Error(detail || `导出失败(${resp.status})`)
  }
  const blob = await resp.blob()
  const name = filenameFrom(resp.headers.get('Content-Disposition'), fallback)
  const href = URL.createObjectURL(blob)
  const a = document.createElement('a')
  a.href = href
  a.download = name
  document.body.appendChild(a)
  a.click()
  a.remove()
  URL.revokeObjectURL(href)
}

export const exportTimetable = (
  semesterId: number, view: ExportView, targetId: number, fmt: ExportFmt,
): Promise<void> =>
  download(
    `/export/timetable?semester_id=${semesterId}&view=${view}&target_id=${targetId}&fmt=${fmt}`,
    `课表.${fmt}`)

export const exportSchoolWorkbook = (semesterId: number): Promise<void> =>
  download(`/export/school.xlsx?semester_id=${semesterId}`, '全校课表总表.xlsx')

export const exportBatchZip = (semesterId: number): Promise<void> =>
  download(`/export/batch.zip?semester_id=${semesterId}`, '全校班级课表.zip')
